### ISSUE TITLE
Reference doc nitpicks

### ISSUE THREAD

**Issue by uecasm** [ned14/outcome#81]
Fixable by me:

- [x] On the landing page there are several undocumented macros listed, which are presumably implementation details.  (Probably already covered by #75.)
- [x] The file content summary at the top of each page includes several detail #include lines that are probably not especially interesting to readers.
- [x] The same also shows some quite lengthy noexcept expressions.  Perhaps this can be elided to just `noexcept(...)` at the top and elaborated in the specific method detail further down?
- [x] It is unclear where `no_exception_type` would be used.  Perhaps this is explained elsewhere?
- [x] In [class outcome](http://ned14.github.io/outcome/reference/outcome/#class-outcome-v2-xxx-outcome) under "Comparison operators" appear two lines `using 'hidden'`, which seem peculiar.  A parsing artifact?
- [x] `outcome` is declared as `outcome<R,S,P>` but its description refers to `outcome<T, E, P>` instead, which might be confusing.  Also neither is actually true since it ignores the `NoValuePolicy`.
- [x] `rebind` does not specify a type for `NoValuePolicy`, which seems like some kind of error; either there's a default parameter which is being hidden or this shouldn't compile.
- [x] The description for `outcome` goes on to mention the policy behaviours in terms of various traits; while the implementation is presumably obvious I couldn't find where `trait::*` are documented (it would be nice if there were links).  Also, it seems perhaps a little naming inconsistency between the `std::` traits that have `_v` suffixes and the `traits::` traits that don't.
- [ ] The description of policy behaviour here seems a bit mixed up between how it selects the default policy based on the type parameters and how a policy (whether selected explicitly or by default) then behaves under particular conditions.  These should probably be split into two separate sections for clarity.
- [x] `NoValuePolicy` has a tiny subheading "Template parameters" by itself, which is followed by larger headings for the other template parameters, which is both out of order and hierarchically confused.
- [x] Static template traits appear to be documented as an "unexposed entity", which seems like a peculiar wording.
- [x] There always seems to be unnecessarily large whitespace between a heading and the code block immediately following it.
- [x] In the [All Narrow](http://ned14.github.io/outcome/reference/policies/all_narrow/) policy, `wide_exception_check` is described as not narrow?
- [ ] Why are `bad_result_access` and `bad_outcome_access` unrelated exceptions?

I didn't have much time to look at it so this was focused mainly on the outcome page and I've only just skimmed the rest.  But hopefully it helps. 😁 

Standardese issues (to be copied to #75)
- [ ] `is_outcome_v` is missing some whitespace (either space or newline) following the template declaration.  This appears to be a general issue for all static or constexpr template values.
- [ ] On the [success/failure page](http://ned14.github.io/outcome/reference/success_failure/), near the documentation for `success_type`/`failure_type` etc, the structure itself and its contained members have the same heading level, which is hierarchically confused and makes reading it harder.  Presumably it makes slightly more sense if you arrive there from an earlier hyperlink and then jump back after reading just one thing, but if you are scrolling down the page it's confusing.
- [ ] The numbering in this section keeps restarting from 1, presumably due to confusion with the nested bullets.

**Comment by ned14** [comment]
Phew. Great help, thanks. I'll start iterating through them, fixing where I can.

**Comment by ned14** [comment]
@foonathan Some of the above apply to Standardese. The rest I'll fix where possible.

**Comment by uecasm** [comment]
For the cases where you've hidden the `noexcept` expression (or actually in general, since even when not hidden they can be hard to parse by humans) it would be useful to add a line to the detailed description; eg for `operator==` something like "_Noexcept:_ if T==T is noexcept" or whatever the condition actually is.

**Comment by ned14** [comment]
Are you not seeing the *Throws:* clause I've painstakingly added to every one of those APIs?

**Comment by uecasm** [comment]
I did overlook that, yes.  At the time I was looking for something that said "noexcept" so it didn't register.

Although a "throws" clause is not quite the same thing; traditionally that's used for "probably will directly throw these exceptions but might indirectly throw something else", whereas noexcept is a stronger guarantee.  It's probably adequate in this case though since `noexcept('hidden')` is still visible on the declaration.

**Comment by ned14** [comment]
Standardese intends to generate reference docs which look like the C++ standard, hence the markup is *Throws:* rather than the noexceptness, which is the inverse. I agree that 'hidden' ought to be "see below" as with the C++ standard.

**Comment by foonathan** [comment]
I did the first round of fixes.

> The file content summary at the top of each page includes several detail #include lines that are probably not especially interesting to readers.

I've converted them into hyperlinks to make them more useful, if they shouldn't be shown, Niall can exclude them.

> In class outcome under "Comparison operators" appear two lines using 'hidden', which seem peculiar. A parsing artifact?

They're gone now.

> while the implementation is presumably obvious I couldn't find where trait::* are documented (it would be nice if there were links).

Could you show me what should be linked to what exactly?

>  On the success/failure page, near the documentation for success_type/failure_type etc, the structure itself and its contained members have the same heading level, which is hierarchically confused and makes reading it harder. Presumably it makes slightly more sense if you arrive there from an earlier hyperlink and then jump back after reading just one thing, but if you are scrolling down the page it's confusing.

Any idea how it could be done better? Currently I'm just generating different heading levels.

**Comment by uecasm** [comment]
> Could you show me what should be linked to what exactly?

Not exactly, because I never did find where they were actually documented.  But the basic idea is that anywhere that a class or method that is implemented within the same body of documentation (though not necessarily the same file) is mentioned, it should be a hyperlink.  I don't know whether it could figure this out automatically like Doxygen or whether it would require an annotation in the source.

> Any idea how it could be done better? Currently I'm just generating different heading levels.

Unless you've changed it since the docs were generated, they're not different levels.  Basically [this](https://ned14.github.io/outcome/reference/success_failure/#struct-outcome-v2-xxx-failure-type) is an h3, and [this](https://ned14.github.io/outcome/reference/success_failure/#type-alias-error-type) is also an h3, despite the latter being a member of the former.  This is confusing.

It does get a little tricky when there's multiple layers of nesting, such as nested namespaces or nested classes, so you wouldn't want to step in a level for all of these.

My best suggestion is that namespaces should all be at one heading level (regardless of nesting depth), structs/classes at the next level (regardless of nesting depth) plus anything else that's at namespace scope, and finally members of structs/classes at the next level.  Additionally, it should try to write all the documentation for the direct members of a class (excluding the nested class itself) before writing any of the documentation for a nested class of that class, to try to keep the grouping more natural and not constantly stepping up and down.

**Comment by ned14** [comment]
I think I've fixed as many of these as I can for now until https://github.com/foonathan/standardese/issues/95 is fixed. Thanks Gavin.

**Comment by ned14** [comment]
Ok, on the remaining items that I can fix:

> The description for outcome goes on to mention the policy behaviours in terms of various traits; while the implementation is presumably obvious I couldn't find where trait::* are documented (it would be nice if there were links). Also, it seems perhaps a little naming inconsistency between the std:: traits that have _v suffixes and the traits:: traits that don't.

I'm blocked on https://github.com/foonathan/standardese/issues/95 for adding links from the reference docs to elsewhere in the docs, but you're right, we'll get those fixed. Does https://ned14.github.io/outcome/tutorial/policies/builtin/ help at all?

The _v vs non-_v is unfortunate. As a C++ 14 library, we must use the non-_v std traits, but as a C++ 20 library, we implement both _v and non-_v traits same as the STL. The same inconsistency pervades the STL nowadays too :(

> The description of policy behaviour here seems a bit mixed up between how it selects the default policy based on the type parameters and how a policy (whether selected explicitly or by default) then behaves under particular conditions. These should probably be split into two separate sections for clarity.

I've completely rewritten outcome's description. See if you like the new version better (the old version was about a year out of date, and totally wrong).

> Why are bad_result_access and bad_outcome_access unrelated exceptions?

Why would they be related? Different types throw different exceptions.

**Comment by uecasm** [comment]
> Does https://ned14.github.io/outcome/tutorial/policies/builtin/ help at all?

Not really; I was talking about the traits, not the policies.

> The _v vs non-_v is unfortunate. As a C++ 14 library, we must use the non-_v std traits, but as a C++ 20 library, we implement both _v and non-_v traits same as the STL. The same inconsistency pervades the STL nowadays too :(

I meant that where the behaviour is described in terms of the traits, you exclusively referred to the std traits with _v and the non-std traits without _v, rather than consistently sticking to one or the other.  (Since both exist, using either in the docs is fine, but it ought to be consistent.)  It looks like that's been fixed in the rewrite though.

> I've completely rewritten outcome's description. See if you like the new version better (the old version was about a year out of date, and totally wrong).

It looks like the second numbered point in the policy discussion [here](https://ned14.github.io/outcome/reference/outcome/#class-outcome-v2-xxx-outcome) has gotten a bit mangled with the "verbatim" leaking through.  And the full justification looks quite bad at medium or less browser widths.

Overall the description does read better than before, but it still looks like it's conflating things.  Maybe I'm just not understanding it properly, but what I was expecting was something more like "if the policy is not specified, and S is void, it defaults to policy::terminate" and basically just list in some kind of table or bullet form how it selects the default policy, without going into detail about what the policies actually do.

And then follow that up with a separate table that says "calling value() erroneously will do X for policy A, Y for policy B, etc" without any mention of what the default is.  Two entirely separate discussions.

Unless it's trying to say that the default policy is not a named one and couldn't be selected explicitly?  If that's the case I would suggest giving it a name (policy::magic maybe) and adding it to the policies chapter like anything else.

> Why would they be related? Different types throw different exceptions.

Result and outcome are both fundamentally the same thing though, they're an encapsulation of a return value and error state, just with differing ideas on the kinds of error state.  As a caller of a method that uses either, I don't want to overly concern myself with which of the two the callee happened to pick, and "bad_result_access" and "bad_outcome_access" both describe pretty much the exact same behaviour, and I don't see making a distinction between them to be useful in any way.  Anyone who wanted to catch one or the other of these would probably want to catch both and do the same action in both cases.

(Somewhat related: I briefly skimmed the docs trying to find a page that might tell me when I should use result vs. outcome.  I didn't find one.)

The OO purist in me gets a little itchy at the other commonalities between result and outcome and wants to give them a common base class or interface as well -- although that's not a serious suggestion; they're templates anyway and C++'s template duck typing lets you get away with a lot of non-pure designs, and in this case it's probably the better choice.

And now for something completely different:

Observe the table of contents [here](https://ned14.github.io/outcome/tutorial/result/).  Now observe it [here](https://ned14.github.io/outcome/tutorial/outcome/).  One of these things is not like the other.

**Comment by ned14** [comment]
I've done some work to remedy the above issues, enough that I consider sufficient for now given the upcoming code freeze. Do however feel free to repeat the bits I haven't addressed during the review, I'll log inidividual issues for each point so they get fixed. Thanks for the feedback Gavin.

**Comment by foonathan** [comment]
I still plan on fixing this one:

> On the success/failure page, near the documentation for success_type/failure_type etc, the structure itself and its contained members have the same heading level, which is hierarchically confused and makes reading it harder. Presumably it makes slightly more sense if you arrive there from an earlier hyperlink and then jump back after reading just one thing, but if you are scrolling down the page it's confusing.

**Comment by foonathan** [comment]
I've fixed that one by showing the full scope name in the heading, this makes it even more obvious than a heading level.
